claim 1-2-3 flow

stenington opened the pull request:
For ticket #36.

stenington committed:
Adding 1-2-3 claim flow with breadcrumbs.
Screen 3 (share) is currently inaccessible because we can't detect when
the user is done sending their badge to the backpack.

stenington committed:
Add simple share page with very boring tweet link

stenington commented on app/templates/core/claim-store.html:
```diff
@@ -17,3 +19,12 @@ <h3 class="badge-name">{{ badge.name }}</h3>
+
+{% block foot %}
+  {{ super() }}
+  <script>
```
This code was a quick solution for the "Send to Backpack" button both opening a new tab and redirecting to the share page, but I just noticed that it doesn't work! In Safari, it's not redirecting. I'll have to fix this.

stenington commented:
This is a first stab at the claim page with the claim-store-share progress indicator thing.

@andrewhayward @christensenep Can one or both of you start to review? I'm happy to move things around if they could be organized better. I wasn't sure exactly how css was organized or which template directories things belonged in, etc.

ghost commented:
This looks pretty good to me!  Your statement about not being sure how the css is organized is very charitable, as it's sort of a total disaster right now, which I take full responsibility for.

The only other organizational issue I can think of is that the templates/core directory could probably just be renamed to "claim" or something, as the only pages that are actually used in there are all claim related (there's still an exports.home in views/core.js that refers to it, but this function is no longer called by anything and is just cruft at this point).  So we could probably just remove that file and rename the directory, if you feel like doing that.

ghost commented:
Btw, let me know if you deploy this, because my PR (https://github.com/mozilla/openbadges-badges/pull/40) is currently unmerged, but its code is actually deployed right now (it allows badges to be issued from the openbadger instance).  It's not a disaster if it is temporarily stomped on, but I want to make sure to stop a related scheduled process first if we do that.

andrewhayward commented on app/templates/core/claim-store.html:
```diff
@@ -17,3 +19,12 @@ <h3 class="badge-name">{{ badge.name }}</h3>
+  {{ super() }}
+  <script>
+    window.onsubmit = function () {
+      window.location = "{{ url('claim.share') }}?email={{ email }}&shortname={{ badge.shortname }}";
```
Is it worth applying your `addQueryString` filter here instead of manually adding these parameters?

stenington replied:
Good catch!

andrewhayward commented on app/index.js:
```diff
@@ -14,6 +14,14 @@ const views = require('./views');
+    if (kwargs[key] !== '' && key !== '__keywords')
+      parts.push(encodeURIComponent(key) + '=' + encodeURIComponent(kwargs[key]));
+  });
+  return url + '?' + parts.join('&');
```
You might want to check that the passed URL doesn't already have a query string here. Also, might be overkill, but is it worth using [the built-in `url` API](http://nodejs.org/api/url.html)?

stenington replied:
Yeah, probably worth it. I took a stab at it and did a little manual testing. The new attempt will override query parameters in the url with parameters in kwargs if the same param is specified in both. I wasn't 100% sure that makes sense, but that's what it does for now.

andrewhayward commented on app/templates/core/claim-breadcrumbs.html:
```diff
@@ -0,0 +1,5 @@
+  <li class="{% if breadcrumb == 'claim' %}current{% else %}unavailable{% endif %}">Claim</li>
+  <li class="{% if breadcrumb == 'store' %}current{% else %}unavailable{% endif %}">Store</li>
+  <li class="{% if breadcrumb == 'share' %}current{% else %}unavailable{% endif %}">Share</li>
+</ul>
```
Would this be better as another macro in `macros/share.html`? `{{ breadcrumbs('claim') }}` seems a bit tidier to me than setting `breadcrumb` and including a snippet.

stenington replied:
Ah, I like that idea.

andrewhayward commented on app/templates/core/claim-share.html:
```diff
@@ -0,0 +1,14 @@
+
+{% block content %}
+{% set breadcrumb = 'share' %}
+{% include "core/claim-breadcrumbs.html" %}
```
See comment above, but for context, this would look a bit tidier as `{{ share.breadcrumbs('share') }}`.

stenington replied:
I may put it elsewhere than `share.html`, just because I can imagine copying the share macros to other projects where we want tweet buttons, etc. Not that they're particularly hard to do, but... might be nice.

andrewhayward replied:
Yeah, put it wherever you think most appropriate. I read `share.html` as being relevant to the 'share' section of the site, rather than for sharing in general. Might be that `breadcrumbs` belongs in `share.html` but `twitterButton` belongs in `utils.html` or something?

andrewhayward commented:
Other than the minor comments, this looks fine to me. Agree with @christensenep about renaming the `core` templates directory - certainly seems like it's worth there being a `claim` directory if we're having multiple `claim` templates.

iamjessklein commented:
+1 to having multiple claim templates - we should explore that further.

stenington committed:
Makes the addQueryString filter more robust and return a SafeString
so the result isn't autoescaped.

Uses addQueryString in send to backpack button redirect, and use
setTimeout which makes Safari happy.

andrewhayward commented on app/index.js:
```diff
@@ -10,10 +10,24 @@ const middleware = require('./middleware');
 env.express(app);
+env.addFilter('addQueryString', function(url, kwargs) {
+  var parseQueryString = true;
+  url = urlUtil.parse(url, parseQueryString);
```
I assume this variable is just here to make it clear what the second parameter is doing? Always torn between doing this and just writing a comment.

stenington replied:
Yeah, maybe it's overkill. I don't like magic boolean arguments, but a comment might be cleaner. I have a tendency not to see comments, so I also tend not to write them.

stenington committed:
Make double-action on form submit work in IE8

stenington committed:
Change commenting of addQueryString

stenington committed:
Move claim breadcrumbs to a macro as per suggestion

stenington commented:
I'll change `core` to `claim` as well, but I'm worried it'll mask the other diffs so I'd rather do it separately.

stenington commented:
Here's the flow as implemented so far:

![screen shot 2013-11-21 at 3 24 55 pm](https://f.cloud.github.com/assets/55319/1595492/38b75678-52eb-11e3-8610-4bb9d7a6a0c6.png)
![screen shot 2013-11-21 at 3 25 08 pm](https://f.cloud.github.com/assets/55319/1595493/3a5b2824-52eb-11e3-91d7-a5e91d9bfbfb.png)
![screen shot 2013-11-21 at 3 25 15 pm](https://f.cloud.github.com/assets/55319/1595494/3bced3b8-52eb-11e3-8651-edb1a9b1c9bd.png)
![screen shot 2013-11-21 at 3 28 58 pm](https://f.cloud.github.com/assets/55319/1595518/91bde642-52eb-11e3-9e4d-ecba800a85b6.png)

I'd prefer to merge it and continue to refine in other PRs. @andrewhayward @christensenep does that work for you?

ghost commented:
Yeah, that makes sense to me.

andrewhayward commented:
Fine with me. I don't see anything worth holding this up for.

stenington committed:
Change claim wording to claim/collect/show

stenington commented:
I changed some words based on a conversation with @xmatthewx, and otherwise we're good to merge this as a first step.

stenington commented:
@christensenep Can you merge this after your PR goes in? Or whenever it's not going to collide with what you're pushing out.

ghost commented:
I'll just merge this right now if that's ok, and then handle the merging with my PR.

ghost commented:
Alright, both this and #40 are now merged.  Happytimes.

iamjessklein commented:
Looking good so far!  I'm eager to see where you get to with the social sharing functionality. 
Is there anything that @xmatthewx and I could do to support you at this point? :ear:

carlacasilli commented:
Hey, I think the terminology is bouncing around here. There is some conversation about "collect" on issue #36

stenington commented:
@carlacasilli If the wording needs to be changed, we can do that. Opening a new ticket would be easiest for me to work with.

threeqube commented:
@stenington we'll resolve the exact wording on the other ticket and open up a new one one to request changes.

iamjessklein commented:
:+1:
